_source/_posts/2021-09-24-java-records.md

```diff
@@ -0,0 +1,556 @@
+---
+layout: blog_post
+title: "Java Records: A WebFlux and Spring Data Example"
+author: jimena-garbarino
+by: contractor
+communities: [java]
+description: "Java Records support in Spring WebFlux Microservices and Spring Data MongoDB Access"
+tags: [java-records, java, api, mongodb, rest, microservices]
+tweets:
+- ""
+- ""
+- ""
+image:
+type: conversion
+---
+When defining classes for a simple aggregation of values, developers had to write constructors, accessors, `equals()`, `hashCode()` and `toString()`, an error-prone ceremony that has low value and deviates the focus from modeling immutable data. With the motivation of simplifying the writing of data carrier classes, Java Records were introduced as a first preview in JDK 14. The second preview came in JDK 15 and the finalized feature came in JDK 16. A summary of the history is available in the [JDK Enhancement Proposal JEP 395](https://openjdk.java.net/jeps/395). While code generators can be used to reduce boilerplate code, the goals of the `record` proposals rather focus on its semantics. In this post let's explore Java Records features and advantages, and apply them for building a REST API and querying a Database.
+
+{% img blog/java-records/openjdk-logo.png alt:"OpenJDK Logo" width:"300" %}{: .center-image }
+
+
+**Prerequisites**:
+
+- [HTTPie](https://httpie.io/)
+- [Java 14+](https://openjdk.java.net/install/index.html)
+- [Okta CLI](https://cli.okta.com)
+
+**Table of Contents**{: .hide }
+* Table of Contents
+{:toc}
+
+# The Record Keyword
+
+The `record` is a new type of declaration, a restricted form of class that acts as a transparent carrier for immutable data. So let's start the exploration by defining a simple data type `EndOfGame` as a `record`:
+
+```java
+import java.time.LocalDate;
+import java.time.LocalTime;
+
+public record EndOfGame(String id, LocalDate date,  LocalTime timeOfDay, String mentalState,
+                        Integer damageTaken, Integer damageToPlayers, Integer damageToStructures) {
+}
+```
+As you can see in the example above, the `record` has a name, `EndOfGame` in the example above. What looks like a constructor signature is the _state description_ or _header_ declaring the _components_ of the `record`.
+
+The following members are acquired automatically with the declaration:
+
+- A private final field and a public read accessor for each component of the state description
+- A public _canonical_ constructor with the same signature as the state description, which initializes each field from the corresponding argument
+- Implementation of `equals()`, `hashCode()` and `toString()`
+
+The code fragments from `EndOfGameTest` class below verify the automatic members are indeed available.
```

Comment:
```suggestion
The code fragments from the `EndOfGameTest` class below verify the automatic members are indeed available.
```

Reply:
The code fragment below looks kinda funny because it's `private` and doesn't seem to be a normal JUnit test.

Reply:
I pasted the whole class and comments are below.